src/sci/impl/reflector.cljc

```diff
@@ -0,0 +1,326 @@
+     [methods args arg-types]
+     (loop [methods methods
+            found-m nil]
+       (if-let [^Method m (first methods)]
+         (let [params (.getParameterTypes m)]
+           (if (and (is-congruent? params args arg-types)
+                    (or (nil? found-m)
+                        (Compiler/subsumes params (.getParameterTypes ^Method found-m))))
+             (recur (rest methods) m)
+             (recur (rest methods) found-m)))
+         found-m))))
+
+#?(:clj
+   (defn- widen-boxed-args
+     "Widen boxed numeric arguments (e.g., Integer -> Long, Float -> Double)."
+     [^objects args]
+     (let [widened (object-array (alength args))]
+       (dotimes [i (alength args)]
+         (let [arg (aget args i)]
+           (if (some? arg)
+             (let [val-class (.getClass ^Object arg)]
+               (aset widened i
+                     (cond
+                       (or (identical? val-class Integer) (identical? val-class Short) (identical? val-class Byte))
+                       (.longValue ^Number arg)
+
+                       (identical? val-class Float)
+                       (.doubleValue ^Number arg)
+
+                       :else
+                       arg)))
+             (aset widened i nil))))
+       widened)))
+
+#?(:clj
+   (defn invoke-matching-method
+     "Invoke a method matching the given name from a list of methods.
+      This is the core SCI-specific method that supports type hints via arg-types.
+      
+      Parameters:
+      - method-name: String name of the method
+      - methods: java.util.List of Method objects
+      - context-class: Class for error messages (can be nil for static methods)
+      - target: Object to invoke on (nil for static methods)
+      - args: Object array of arguments
+      - arg-types: Optional array of Class objects for type hints"
+     ([method-name methods ^Object target args]
+      (invoke-matching-method method-name methods
+                              (when target (.getClass target))
+                              target args nil))
+     ([method-name methods context-class target args]
+      (invoke-matching-method method-name methods context-class target args nil))
+     ([method-name methods context-class target args arg-types]
+      (let [methods (seq methods)]
+        (if (empty? methods)
+          (throw (IllegalArgumentException.
+                  (str "No matching method " method-name " found taking "
+                       (alength ^objects args) " args"
+                       (when context-class (str " for " context-class)))))
+          (let [^Method m (if (= 1 (count methods))
+                            (first methods)
+                            (or (match-method methods args arg-types)
+                                ;; widen boxed args and re-try
+                                (match-method methods (widen-boxed-args args) arg-types)))]
```

Comment:
BUG! Original code updates `args` for use later in the method:

```
			args = widenBoxedArgs(args);
			m = matchMethod(methods, args, argTypes);
```

Reply:
This actually isn't a bug since widen-boxed-args also updates the args and returns it

Reply:
I'm not sure if this will really result in a bug though. It'd be interested to extract a test case for this one to see where it currently fails.

Reply:
Since the args are widened to find the correct method, e.g. Double -> whatever instead of Float -> whatever. But then later the Float argument is boxed and passed to the method which may just work...?

Reply:
Here is an example that this will just work out ok I think:

``` clojure
(ns reflect-example
  (:import [java.lang Math]
           [java.lang.reflect Method]))

;; Get the Method object for Math.pow(double, double)
(def m (.getMethod Math "pow" (into-array Class [Double/TYPE Double/TYPE])))

;; Float value
(def f (float 3.0))
(prn (type f))

;; Attempt to invoke with Float (boxed) arguments
(try
  (.invoke m nil (into-array Object [f (float 1.0)]))
  (catch IllegalArgumentException e
    (println "Caught exception:" (.getMessage e))))
```

Reply:
Here we can see that it works in bb:

```
$ ./bb -e '(Math/pow (float 1.33) (float 2.0))'
1.7689001141548175
```

Reply:
I'll ask in clojure-dev. We could make widen-box-args mutate the original array if we want re-assigment.